File: utils/password.go
// Copyright (c) 2016-present Mattermost, Inc. All Rights Reserved.
// See License.txt for license information.

package utils

import (
	"net/http"
	"strings"

	"github.com/mattermost/mattermost-server/model"
)

func IsPasswordValid(password string) *model.AppError {
	if len(password) > model.PASSWORD_MAXIMUM_LENGTH || len(password) < model.PASSWORD_MINIMUM_LENGTH {
		return model.NewAppError("User.IsValid", "model.user.is_valid.pwd.app_error", map[string]interface{}{"Min": model.PASSWORD_MINIMUM_LENGTH}, "", http.StatusBadRequest)
	}

	return nil
}

func IsPasswordValidWithSettings(password string, settings *model.PasswordSettings) *model.AppError {
	id := "model.user.is_valid.pwd"
	isError := false

	if len(password) < *settings.MinimumLength || len(password) > model.PASSWORD_MAXIMUM_LENGTH {
		isError = true
	}

	if *settings.Lowercase {
		if !strings.ContainsAny(password, model.LOWERCASE_LETTERS) {
			isError = true
		}

		id = id + "_lowercase"
	}

	if *settings.Uppercase {
		if !strings.ContainsAny(password, model.UPPERCASE_LETTERS) {
			isError = true
		}

		id = id + "_uppercase"
	}

	if *settings.Number {
		if !strings.ContainsAny(password, model.NUMBERS) {
			isError = true
		}

		id = id + "_number"
	}

	if *settings.Symbol {
		if !strings.ContainsAny(password, model.SYMBOLS) {
			isError = true
		}

		id = id + "_symbol"
	}

	if isError {
		return model.NewAppError("User.IsValid", id+".app_error", map[string]interface{}{"Min": *settings.MinimumLength}, "", http.StatusBadRequest)
	}

	return nil
}

File: e2e-tests/playwright/lib/src/ui/components/system_console/sections/authentication/password.ts
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {Locator} from '@playwright/test';
import {expect} from '@playwright/test';

/**
 * System Console -> Authentication -> Password
 */
export default class PasswordSettings {
    readonly container: Locator;

    readonly minimumLength: Locator;
    readonly maximumLoginAttempts: Locator;
    readonly lowercase: Locator;
    readonly uppercase: Locator;
    readonly number: Locator;
    readonly symbol: Locator;
    readonly saveButton: Locator;
    readonly lengthError: Locator;
    readonly sampleHelpText: Locator;

    constructor(container: Locator) {
        this.container = container;

        this.minimumLength = container.getByPlaceholder('E.g.: "5"');
        this.maximumLoginAttempts = container.getByPlaceholder('E.g.: "10"');
        this.lowercase = container.getByRole('checkbox', {name: 'At least one lowercase letter'});
        this.uppercase = container.getByRole('checkbox', {name: 'At least one uppercase letter'});
        this.number = container.getByRole('checkbox', {name: 'At least one number'});
        this.symbol = container.getByRole('checkbox', {name: 'At least one symbol (e.g. "~!@#$%^&*()")'});
        this.saveButton = container.getByRole('button', {name: 'Save'});
        this.lengthError = container.getByText(
            /Minimum password length must be a whole number greater than or equal to (5|14) and less than or equal to 72\./,
        );
        this.sampleHelpText = container.getByText(/Your password must be \d+-72 characters long/);
    }

    async toBeVisible() {
        await expect(this.minimumLength).toBeVisible();
    }

    async save() {
        await expect(this.saveButton).toBeEnabled();
        await this.saveButton.click();
    }

    async reload() {
        await this.container.page().reload();
        await this.toBeVisible();
    }
}
